addOns/websocket/src/main/java/org/zaproxy/zap/extension/websocket/pscan/WebSocketPassiveScanThread.java

```diff
@@ -0,0 +1,189 @@
+    
+    /**
+     * Thread safe Linked Queue
+     */
+    private LinkedBlockingQueue<MessageWrapper> messagesBuffer;
+    
+    /**
+     * {@code True} to enable the passive scan Thread
+     */
+    private volatile boolean isPassiveScannerActive;
+    
+    /**
+     * Reference to Database. Used in order to pick messages
+     * for scanning
+     */
+    private TableWebSocket tableWebSocket;
+    
+    /**
+     * Manager used to updating the messages table. In addition,
+     * used by passive scan thread so as to get access at plugins
+     */
+    private WebSocketPassiveScannerManager passiveScannerManager;
+    
+    /**
+     * Initialize the passive scan in background thread. By default thread is
+     * inactive and not alive. In order to activate thread set
+     * {@link WebSocketPassiveScanThread#setPassiveScannerActive(boolean)} true and
+     * start thread with {@link WebSocketPassiveScanThread#start()}.
+     * @param passiveScannerManager the manager
+     */
+    public WebSocketPassiveScanThread (WebSocketPassiveScannerManager passiveScannerManager) {
+        this.passiveScannerManager = passiveScannerManager;
+        this.isPassiveScannerActive = false;
+        messagesBuffer = new LinkedBlockingQueue<>();
+    }
+    
+    /**
+     * @return true if the table was initialized
+     */
+    public boolean hasTableWebSocket(){
+        return tableWebSocket != null;
+    }
+    
+    public void setTableWebSocket(TableWebSocket tableWebSocket) {
+        this.tableWebSocket = tableWebSocket;
+    }
+    
+    public TableWebSocket getTableWebSocket() {
+        return tableWebSocket;
+    }
+    
+    @Override
+    public int getListenerOrder() {
+        return WEBSOCKET_LISTENER_ORDER;
+    }
+    
+    @Override
+    public void onMessageFrame(int channelId, WebSocketMessage message, WebSocketProxy.Initiator initiator) {
+        //TODO: Take care {@link WebSocketMessage#OPCODE_CONTINUATION}
+        messagesBuffer.add(new MessageWrapper(message.getMessageId(),channelId,initiator));
+    }
+    
+    public boolean isPassiveScannerActive(){
+        return isPassiveScannerActive;
+    }
+    
+    public void setPassiveScannerActive(boolean passiveScannerActive) {
+        isPassiveScannerActive = passiveScannerActive;
+    }
+    
+    @Override
+    public void onStateChange(WebSocketProxy.State state, WebSocketProxy proxy) {
+        //Ignore
+    }
+    
+    @Override
+    public void run() {
+        MessageWrapper messageWrap;
+        WebSocketPassiveScannerPlugin currentPassiveScanner;
+        WebSocketMessageDTO currentMessage;
+        Iterator<WebSocketPassiveScanner> iterator;
+        
+        while (isPassiveScannerActive){
+            if (messagesBuffer.isEmpty() || tableWebSocket == null) {
+                try {
+                    Thread.sleep(SLEEP_TIME);
+                } catch (InterruptedException e) {
+                    LOGGER.info("Sleeping was interrupted",e);
```

Comment:
Should this be moved to a different level? When would this be encountered?

Reply:
This is expected not to be triggered ever.  Even if that happened,  it will not indicate error or warning. So this looks good to me.